x/psiphon/psiphon.go

```diff
@@ -208,21 +88,51 @@ func (d *Dialer) DialStream(unusedContext context.Context, addr string) (transpo
 	return streamConn{netConn}, nil
 }
 
-// Start configures and runs the Dialer. It must be called before you can use the Dialer. It returns when the tunnel is ready.
-func (d *Dialer) Start(ctx context.Context, config *Config) error {
-	configJSON, err := json.Marshal(config)
-	if err != nil {
-		return fmt.Errorf("failed to convert config to JSON")
+func newPsiphonConfig(config *DialerConfig) (*psi.Config, error) {
+	if config == nil {
+		return nil, errors.New("config must not be nil")
 	}
-	pConfig, err := psi.LoadConfig(configJSON)
+	// Validate keys. We parse as a map first because we need to check for the existence
+	// of certain keys.
+	var configMap map[string]interface{}
+	if err := json.Unmarshal(config.ProviderConfig, &configMap); err != nil {
+		return nil, fmt.Errorf("failed to parse config: %w", err)
+	}
+	for key, value := range configMap {
+		switch key {
+		case "DisableLocalHTTPProxy", "DisableLocalSocksProxy":
+			b, ok := value.(bool)
+			if !ok {
+				return nil, fmt.Errorf("field %v must be a boolean", key)
+			}
+			if b != true {
+				return nil, fmt.Errorf("field %v must be true if set", key)
+			}
+		case "DataRootDirectory":
```

Comment:
Why don't we allow customers to specify their own `DataRootDirectory`?

Reply:
Because that must be specified via the DialerConfig.DataRootDirectory. This makes it clear what they should be setting vs what is provided by Psiphon.